workers/workers_test.go

```diff
@@ -131,3 +135,180 @@ func BenchmarkSerial(b *testing.B) {
 		})
 	}
 }
+
+func TestJobWait(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+	testError := errors.New("TestError")
+	go func() {
+		job.result <- testError
+	}()
+	returned := job.Wait()
+	require.ErrorIs(testError, returned, "Incorrect error returned.")
+}
+
+func TestJobDone(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+	var callMU sync.Mutex
+	called := false
+	job.Done(func() {
+		callMU.Lock()
+		defer callMU.Unlock()
+		called = true
+	})
+	job.completed <- struct{}{}
+	// Check that the callback function was called
+	callMU.Lock()
+	require.True(called, "Callback function not called")
+	callMU.Unlock()
+
+	// Ensure task channel was closed
+	require.Empty(job.tasks, "Tasks channel not closed")
+}
+
+func TestJobGo(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error, 2),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+	n := 1
+	testError := errors.New("TestError")
+	for n <= 2 {
+		job.Go(func() error {
+			if n == 1 {
+				return nil
+			} else {
+				return testError
+			}
+		})
+		n += 1
+	}
+	n = 1
+	// Ensure corret errors
+	for n <= 2 {
+		response := <-job.tasks
+		if n == 1 {
+			require.NoError(response(), "Incorrect error message.")
+		} else {
+			require.ErrorIs(testError, response(), "Incorrect error message.")
+		}
+		n += 1
+	}
+}
+
+func TestStopWorker(t *testing.T) {
+	require := require.New(t)
+	w := New(10, 100)
+	require.False(w.shouldShutdown, "Shutdown val incorrectly initialized.")
+	w.Stop()
+	require.Empty(w.queue, "Worker queue not empty")
+	require.True(w.shouldShutdown, "Shutdown val not set.")
+}
+
+func TestWorker(t *testing.T) {
+	require := require.New(t)
+	w := New(10, 10000)
+	// Require workers was created properly
+	require.Equal(10, w.count, "Count not set correctly")
+	require.Empty(w.queue, "Worker queue not empty")
+	require.Empty(w.tasks, "Worker tasks not empty")
+	// Shutdown fields
+	require.Empty(w.ackShutdown, "Worker ackShutdown not empty")
+	require.Empty(w.stopWorkers, "Worker stopWorkers not empty")
+	require.Empty(w.stoppedWorkers, "Worker stoppedWorkers not empty")
+	// Wait on jobs
+	var wg sync.WaitGroup
+	wg.Add(1000)
+	// Value updated by the workers
+	var valLock sync.Mutex
+	val := 0
+	for i := 0; i < 1000; i++ {
+		// Create a new job
+		job, err := w.NewJob(10)
+		require.NoError(err)
+		for j := 0; j < 10; j++ {
+			task := func() error {
+				valLock.Lock()
+				defer valLock.Unlock()
+				val += 1
+				return nil
+			}
+			job.Go(task)
+		}
+		go func() {
+			defer wg.Done()
+			err := job.Wait()
+			require.NoError(err, "Error waiting on job.")
+		}()
+		job.Done(nil)
+	}
+	finished := make(chan bool)
```

Comment:
You should enqueue the jobs you created and just wait on each one out here

Reply:
something like:
```
for _, j := range jobs {
  require.NoError(j.Wait(), "Error waiting on job.")
}
```

Reply:
You can then get rid of all this "finish" logic

you can call `w.Stop` after ensuring all jobs finished